presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java

```diff
@@ -507,49 +516,53 @@ private static void logQueryTimeline(
                 queryEndTime);
     }
 
-    private static List<StageCpuDistribution> getCpuDistributions(QueryInfo queryInfo)
-    {
-        if (!queryInfo.getOutputStage().isPresent()) {
-            return ImmutableList.of();
-        }
-
-        ImmutableList.Builder<StageCpuDistribution> builder = ImmutableList.builder();
-        populateDistribution(queryInfo.getOutputStage().get(), builder);
-
-        return builder.build();
-    }
-
-    private static void populateDistribution(StageInfo stageInfo, ImmutableList.Builder<StageCpuDistribution> distributions)
-    {
-        distributions.add(computeCpuDistribution(stageInfo));
-        for (StageInfo subStage : stageInfo.getSubStages()) {
-            populateDistribution(subStage, distributions);
-        }
-    }
-
-    private static StageCpuDistribution computeCpuDistribution(StageInfo stageInfo)
+    private static void computeTaskDistributions(
+            StageInfo stageInfo,
+            ImmutableList.Builder<TaskDistribution> cpuDistributionBuilder,
+            ImmutableList.Builder<TaskDistribution> memoryDistributionBuilder)
     {
         Distribution cpuDistribution = new Distribution();
+        Distribution memoryDistribution = new Distribution();
 
         for (TaskInfo taskInfo : stageInfo.getTasks()) {
             cpuDistribution.add(taskInfo.getStats().getTotalCpuTime().toMillis());
+            memoryDistribution.add(taskInfo.getStats().getPeakTotalMemoryInBytes());
         }
 
-        DistributionSnapshot snapshot = cpuDistribution.snapshot();
+        DistributionSnapshot cpuSnapshot = cpuDistribution.snapshot();
+        DistributionSnapshot memorySnapshot = memoryDistribution.snapshot();
 
-        return new StageCpuDistribution(
+        cpuDistributionBuilder.add(new TaskDistribution(
                 stageInfo.getStageId().getId(),
                 stageInfo.getTasks().size(),
-                snapshot.getP25(),
-                snapshot.getP50(),
-                snapshot.getP75(),
-                snapshot.getP90(),
-                snapshot.getP95(),
-                snapshot.getP99(),
-                snapshot.getMin(),
-                snapshot.getMax(),
-                (long) snapshot.getTotal(),
-                snapshot.getTotal() / snapshot.getCount());
+                cpuSnapshot.getP25(),
```

Comment:
Constructing a task distribution is very verbose and a lot of this stuff comes from a DistributionSnapshot, can you update the TaskDistributionConstructor to reduce some of this duplication by working on the DistributionSnapshot directly?

Reply:
This is intended to support serialization into json string. There is downstream dependency on structure of class.